testng-core/src/test/java/test/timeout/github2672/TimeoutStacktraceTest.java

```diff
@@ -0,0 +1,41 @@
+package test.timeout.github2672;
+
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.util.Arrays;
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+import org.testng.internal.thread.ThreadTimeoutException;
+
+@Listeners(TimeoutStacktraceTest.class)
```

Comment:
You don't want to add a listener on the TestNG test but on the TestNG sample.

Move your test sample into a `*Sample` class.
From a `*Test` build your TestNG run (add your listener instance + sample) and assert there.